### PR TITLE
Rh photogram industrial 2

### PR DESCRIPTION
**Pull request by rezahusain**

### CODE REVIEW

**Review by aborov — APPROVED** [review]
Looks good!

**Comment by thierrychau** [comment]
> Looks good!

Aleksei, I appreciate you helping you fellow cohort-mates so much with providing many code reviews. However, I want to remind you that each one of you will need to provide a code review per PR, please leave some for your peers :)

File: app/controllers/application_controller.rb
Comment (thierrychau) [inline app/controllers/application_controller.rb:1]: I don't think these adds should be in this pull request. 

It's good practice to keep pull requests very focused on what is asked of you, especially when contributing to larger code where other changes may need some other approval. 

In addition, keeping pull request focused on one feature helps tremendously with code review. Reviewing a pull request involving 78 file changes is complicated enough, no need to more noise to it.

I do see that you are doing more, and very possibly better. In another scenario I'd probably let this pass. In this case, for example `authenticate_user!` is part of the next pull request. If I let it pass now, it's more work down the road for next reviewers to figure out why this doesn't show up in the next pull request.

Can you remove this for now?

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class UsersController < ApplicationController
+  def show
+    @user = current_user
+  end
```
Comment (thierrychau) [inline app/controllers/users_controller.rb]: We do not want this here, yet. Please remove.

File: app/models/like.rb [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#  index_likes_on_photo_id  (photo_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (fan_id => users.id)
+#  fk_rails_...  (photo_id => photos.id)
+#
+class Like < ApplicationRecord
+  belongs_to :fan, class_name: "User", counter_cache: true
+  belongs_to :photo, counter_cache: true
```
Comment (thierrychau) [inline app/models/like.rb:23]: Can you add a validation to make sure a user can only like a photo once, and no more than that?

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+
+  has_many :own_photos, class_name: "Photo", foreign_key: "owner_id"
+
+  has_many :comments, foreign_key: "author_id"
+
+  has_many :sent_follow_requests, foreign_key: :sender_id, class_name: "FollowRequest"
```
Comment (thierrychau) [inline app/models/user.rb]: Can you add an option to prevent getting an error for the foreign key constraint if a user gets deleted and leaves associated records with a user no longer existing?

File: app/views/assets/_navbar.html.erb
Comment (thierrychau) [inline app/views/assets/_navbar.html.erb:1]: I appreciate the effort to go above and beyond what is asked of you, however this should not be part of this pull request. We want to keep you focused on what is asked of you.

File: app/views/devise/mailer/email_changed.html.erb
Comment (thierrychau) [inline app/views/devise/mailer/email_changed.html.erb:1]: We do not want all of the devise's views for now, please remove them.

File: app/views/layouts/application.html.erb
Comment (thierrychau) [inline app/views/layouts/application.html.erb:1]: We do not want these, yet.

File: config/initializers/devise.rb
Comment (thierrychau) [inline config/initializers/devise.rb:1]: Comment out that line

File: config/routes.rb [hunk 4]
```diff
@@ -1,7 +1,13 @@
...
+  resources :follow_requests
+  resources :comments
+  resources :photos
+  devise_for :users
   # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
 
   # Defines the root path route ("/")
   # root "articles#index"
   root "photos#index"
+  get '/:username' => 'users#show'
```
Comment (thierrychau) [inline config/routes.rb]: We don't want this, yet.

File: public/dev.rake
Comment (thierrychau) [inline public/dev.rake:1]: Delete this file

**Review by thierrychau — CHANGES_REQUESTED** [review]
Hi Reza, great job. I see that you are putting in extra work and effort. While it would usually be commendable, here we want to keep you very focused and concise. Please see below comments, happy to discuss these tomorrow!

File: app/controllers/application_controller.rb
Comment (rezahusain) [reply app/controllers/application_controller.rb:1]: Hello Thierry,

I was able to resolve almost all of the existing code. Really sorry about the extra stuff, over the weekend I saw that the first part was worth 10 points and was working towards that until I realized that we only had to get devise set up. 

Just a quick question - For the validations are you thinking about an upcoming section? I went through the lecture notes and video and he never mentioned validating the amount of times a user likes or regarding foreign constraints.

Reza

Comment (thierrychau) [reply app/controllers/application_controller.rb:1]: Hi Reza, no need to apologize. This is great work on your part. 

Normally we would have more back and forth and debate whether each great addition you have should be here or somewhere else. We would probably have separated this one PR into several (probably one per model), and everything you have thought of would probably belong somewhere in each of them. Here the difficulty is that we have limited time, and we have structured PRs going one specific direction.

As for the foreign key constraint, I was referring to this paragraph in part 2:

> Also, the order of .destroy_all matters! What if we put User.destroy_all above FollowRequest.destroy_all. What would happen?
>We would get a violation of the foreign key (fk) constraint. That’s because the FollowRequests contain IDs from the User, and the references method we used in our migrations creates a constraint whereby you can’t delete a parent object (User) before deleting the children (FollowRequests).
>Alternatively, this has shown us that we may want to add another option onto the has_manys associated with follow requests: dependent: :destroy. For instance, adding that option to sent_follow_requests will cause FollowRequest records to be deleted when a User is deleted. Probably a good idea, but I leave that up to you to add!

Comment (thierrychau) [reply app/controllers/application_controller.rb:1]: I think these changes are still present

**Review by thierrychau — APPROVED** [review]
Good to merge, great job Reza!
